### PR TITLE
TPD-1154 Process Gift API

### PR DESCRIPTION
**Pull request by jason-rice**
[touchpointsoftware.atlassian.net/browse/TPD-1154](https://touchpointsoftware.atlassian.net/browse/TPD-1154)

### CODE REVIEW

File: UnitTests/CMSWebTests/Areas/Finance/GatewayTests.cs [hunk 1]
```diff
@@ -3,110 +3,317 @@
...
-                ("PublicSalt", "66 82 79 78 66 82 79 78")
+                ("PublicSalt", "66 82 79 78 66 82 79 78"),
+                ("sysfromemail", "info@touchpointsoftware.com")
             );
+            var contributionFund = MockFunds.CreateContributionFund(db, null);
+            var pageName = "Giving Page Test, " + DateTime.Now.ToString();
+            newGivingPage = MockGivingPage.CreateGivingPage(db, pageName, contributionFund.FundId, 1);
         }
 
+        //[Fact]
```
Comment (timdavila) [inline UnitTests/CMSWebTests/Areas/Finance/GatewayTests.cs:33]: remove this please

File: CmsData/Finance/TransNationalGateway.cs [hunk 2]
```diff
@@ -93,8 +99,106 @@ public TransactionResponse AuthCreditCard(int peopleId, decimal amt, string card
...
+            };
+        }
+
+        public TransactionResponse ChargeBankAccountOneTime(decimal amt, string accountNumber, string routingNumber, string accountName, string nameOnAccount, string description, string firstName, string lastName, string address, string address2, string city, string state, string country, string zip, string phone, string email, bool testing = false)
+        {
+            const SslProtocols _Tls12 = (SslProtocols)0x00000C00;
+            const SecurityProtocolType Tls12 = (SecurityProtocolType)_Tls12;
+            ServicePointManager.SecurityProtocol = Tls12;
+
+            //var type = AchType(peopleId);
```
[the remark this answers is not in the document]
Comment (jason-rice) [reply CmsData/Finance/TransNationalGateway.cs:149]: resolved

File: CmsData/Finance/TransNationalGateway.cs [hunk 3]
```diff
@@ -93,8 +99,106 @@ public TransactionResponse AuthCreditCard(int peopleId, decimal amt, string card
...
+            };
+            var achSaleRequest = new AchSaleRequest(
+                _userName,
+                _password,
+                ach,
+                amt);
+
+            var response = achSaleRequest.Execute();
+
+            if (type == "savings")
```
[the remark this answers is not in the document]
Comment (jason-rice) [reply CmsData/Finance/TransNationalGateway.cs:179]: resolved

File: CmsWeb/Areas/Giving/Models/GivingPaymentModel.cs [hunk 4]
```diff
@@ -27,31 +31,84 @@ public Message CreateMethod(GivingPaymentViewModel viewModel)
...
-                    return Message.createErrorReturn("State required.", Message.API_ERROR_PAYMENT_METHOD_REQUIRED_FIELD_EMPTY);
-                if (viewModel.country == "")
-                    return Message.createErrorReturn("Country required.", Message.API_ERROR_PAYMENT_METHOD_REQUIRED_FIELD_EMPTY);
-                if (viewModel.zip == "")
-                    return Message.createErrorReturn("Zip required.", Message.API_ERROR_PAYMENT_METHOD_REQUIRED_FIELD_EMPTY);
-            }
-            if (viewModel.transactionTypeId == "")
-                return Message.createErrorReturn("Transaction Type ID required.", Message.API_ERROR_PAYMENT_METHOD_REQUIRED_FIELD_EMPTY);
+            }
+            if (viewModel.paymentTypeId == 1)
```
[the remark this answers is not in the document]
Comment (jason-rice) [reply CmsWeb/Areas/Giving/Models/GivingPaymentModel.cs:78]: fixed this, great catch!

File: CmsWeb/Areas/Giving/Models/GivingPaymentModel.cs [hunk 5]
```diff
@@ -27,31 +31,84 @@ public Message CreateMethod(GivingPaymentViewModel viewModel)
...
                 if (viewModel.peopleId == null)
+                {
                     return Message.createErrorReturn("People Id is a required field, but is missing.", Message.API_ERROR_SCHEDULED_GIFT__PEOPLE_ID_NOT_FOUND);
+                }
+            }
+            if (viewModel.billingInfo == null)
+            {
+                return Message.createErrorReturn("Billing info object is required.", Message.API_ERROR_PAYMENT_METHOD_REQUIRED_FIELD_EMPTY);
+            }
+            if (viewModel.billingInfo.firstName == "")
```
[the remark this answers is not in the document]
Comment (jason-rice) [reply CmsWeb/Areas/Giving/Models/GivingPaymentModel.cs:46]: changed these

File: CmsWeb/Areas/Giving/Models/GivingPaymentModel.cs [hunk 6]
```diff
@@ -60,177 +117,54 @@ public Message CreateMethod(GivingPaymentViewModel viewModel)
...
                 currentPeopleId = (int)CurrentDatabase.UserPeopleId;
+            }
             else
+            {
                 currentPeopleId = (int)viewModel.incomingPeopleId;
+            }
 
-            if (viewModel.testing == true)
-                switch (viewModel.paymentTypeId)
+            if (viewModel.paymentTypeId == 1)
```
[the remark this answers is not in the document]
Comment (jason-rice) [reply CmsWeb/Areas/Giving/Models/GivingPaymentModel.cs:128]: fixed

File: CmsWeb/Areas/Giving/Models/GivingPaymentModel.cs [hunk 7]
```diff
@@ -530,7 +466,303 @@ public Message DeleteSchedule(Guid? scheduledGiftId, int? peopleId = null)
...
+        {
+            // checking for all required inputs
+            #region
+            if (viewModel.testing == false)
+            {
+                if (viewModel.peopleId != CurrentDatabase.CurrentPeopleId)
+                {
+                    return Message.createErrorReturn("People Id does not match the current people id.", Message.API_ERROR_SCHEDULED_GIFT__PEOPLE_ID_NOT_FOUND);
+                }
+                if (viewModel.peopleId == null)
```
[the remark this answers is not in the document]
Comment (jason-rice) [reply CmsWeb/Areas/Giving/Models/GivingPaymentModel.cs:482]: good catch! fixed

File: UnitTests/CMSWebTests/Areas/Finance/GatewayTests.cs (hunk 1 above)
Comment (jason-rice) [reply UnitTests/CMSWebTests/Areas/Finance/GatewayTests.cs:33]: fixed

File: CmsWeb/Areas/Giving/Models/GivingPaymentModel.cs [hunk 8]
```diff
@@ -530,7 +466,303 @@ public Message DeleteSchedule(Guid? scheduledGiftId, int? peopleId = null)
...
+                    }
+                    if (viewModel.testing == false)
+                    {
+                        var emailFromList = givingPage.OnlineNotifyPerson?.Split(',');
+                        List<MailAddress> mailAddressesTO = new List<MailAddress>();
+                        List<MailAddress> mailAddressesCC = new List<MailAddress>();
+                        mailAddressesTO.Add(new MailAddress(viewModel.billingInfo.email));
+                        if (emailFromList == null)
+                        {
+                            return Message.successMessage("Credit card payment processed successfully. Confirmation email failed. Message: " + transactionResponseForBankPayment.Message, Message.API_ERROR_CONFIRMATION_EMAIL_FAILED, totalAmount);
```
[the remark this answers is not in the document]
Comment (jason-rice) [reply CmsWeb/Areas/Giving/Models/GivingPaymentModel.cs:664]: fixed this

File: CmsWeb/Areas/Giving/Models/GivingPaymentModel.cs [hunk 9]
```diff
@@ -530,7 +466,303 @@ public Message DeleteSchedule(Guid? scheduledGiftId, int? peopleId = null)
...
+
+                if (transactionResponseForBankPayment.Approved == true)
+                {
+                    foreach (var item in viewModel.gifts)
+                    {
+                        var contribution = new Contribution
+                        {
+                            CreatedBy = 0,
+                            CreatedDate = DateTime.Now,
+                            ContributionTypeId = 5,
```
[the remark this answers is not in the document]
Comment (jason-rice) [reply CmsWeb/Areas/Giving/Models/GivingPaymentModel.cs:630]: fixed

File: CmsWeb/Areas/Giving/Models/GivingPaymentModel.cs [hunk 10]
```diff
@@ -530,7 +466,303 @@ public Message DeleteSchedule(Guid? scheduledGiftId, int? peopleId = null)
...
+                            ContributionDate = DateTime.Now,
+                            ContributionAmount = item.amount,
+                            ImageID = 0,
+                            Origin = 0
```
[the remark this answers is not in the document]
Comment (jason-rice) [reply CmsWeb/Areas/Giving/Models/GivingPaymentModel.cs:634]: fixed

File: CmsWeb/Areas/Giving/Models/GivingPaymentModel.cs [hunk 11]
```diff
@@ -60,177 +117,54 @@ public Message CreateMethod(GivingPaymentViewModel viewModel)
...
-                    case 99: // Other
-                        break;
-                    default:
-                        break;
+                    PeopleId = currentPeopleId,
+                    PaymentMethodTypeId = (int)viewModel.paymentTypeId,
+                    IsDefault = viewModel.isDefault,
+                    Name = viewModel.bankInfo.accountName,
+                    NameOnAccount = viewModel.billingInfo.firstName + " " + viewModel.billingInfo.lastName,
+                    MaskedDisplay = "•••• •••• •••• 1234",
```
[the remark this answers is not in the document]
Comment (jason-rice) [reply CmsWeb/Areas/Giving/Models/GivingPaymentModel.cs:137]: added GivingPaymentViewModel.maskedDisplay

**Pull request merged**